working/unquoted-imports/feature-specification.md

```diff
@@ -291,27 +355,33 @@ We add a new rule and hang it off the existing `uri` rule already used by import
 and export directives:
 
 ```
-uri                   ::= stringLiteral | packagePath
-packagePath           ::= packagePathSegment ( '/' packagePathSegment )*
-packagePathSegment    ::= dottedIdentifierList
-dottedIdentifierList  ::= identifier ('.' identifier)*
+uri               ::= stringLiteral | packagePath
+packagePath       ::= pathSegment ( '/' pathSegment )*
+pathSegment       ::= segmentComponent ( '.' segmentComponent )*
+segmentComponent  ::= identifier
+                    | ⟨RESERVED_WORD⟩
+                    | ⟨BUILT_IN_IDENTIFIER⟩
+                    | ⟨OTHER_IDENTIFIER⟩
 ```
 
-An import or export can continue to use a `stringLiteral` for the quoted form
-(which is what they will do for relative imports). But they can also use a
-`packagePath`, which is a slash-separated series of segments, each of which is a
-series of dot-separated identifiers. *(The `dottedIdentifierList` rule is
-already in the grammar and is shown here for clarity.)*
+It is a compile-time error if any whitespace, newlines, or comments occur
+between any of the `segmentComponent`, `/`, or `.` tokens in a `packagePath`.
+*In other words, there can be nothing except the terminals themselves from the
+first `segmentComponent` in the `packagePath` to the last.*
+
+*An import, export, or part directive can continue to use a `stringLiteral` for
+the quoted form (which is what they will do for relative references). But they
+can also use a `packagePath`, which is a slash-separated series of segments,
+each of which is a series of dot-separated components.*
 
 ### Part directive lookahead
 
-*There are two directives for working with part files, `part` and `part of`. The
-`of` identifier is not a reserved word in Dart. This means that when the parser
-sees `part of`, it doesn't immediately know if it is looking at a `part`
-directive followed by an unquoted identifier like `part of;` or `part
-of.some/other.thing;` versus a `part of` directive like `part of thing;` or
-`part of 'uri.dart';` It must lookahead past the `of` identifier to see if the
-next token is `;`, `.`, `/`, or another identifier.*
+*There are two directives for working with part files, `part` and `part of`.
+This means that when the parser sees `part of`, it doesn't immediately know if
+it is looking at a `part` directive followed by an unquoted identifier like
+`part of;` or `part of.some/other.thing;` versus a `part of` directive like
+`part of thing;` or `part of 'uri.dart';` It must lookahead past the `of`
+identifier to see if the next token is `;`, `.`, `/`, or another identifier.*
```

Comment:
> If we allow that, then `part of deferred as f;` would be ambiguous, if seen by itself.

So you're saying this could be either:

* A `part of` directive for a file named `deferred` which should be locally prefixed as `f`.
* A `part` directive for a file named `of` which is deferred loaded and available under prefix `f`.

Is that right? If so, I don't think it's ambiguous since the latter isn't valid. We only allowed `deferred` on imports, not `part`, right? Does enhanced parts change that?

Reply:
That was what I thought when I wrote it, but you can't have deferred parts (well, yet! ... but likely ever) so that's not really ambiguous.

So what I should have said is: If we allow *something* after `part <uri>` in the future, and that *something* can be a single identifier or dotted identifier, then <code>part of *something*;</code> is ambiguous.

And if we allow things after both `part <uri>` and `part of <uri>`, then if any *thing* after `part <uri>` has the form `validUnquotedUri otherThing` where *otherThing` is valid after `part of <uri>`,  then `part of *thing*` is ambiguous.

A lot of "if"s.

The things I have actually considered so far are, during "parts with modifiers" discussions:
* `as id` modifiers on `part of` and `part` to contain the identifiers they introduce.
* `sealed` modifier on `part of` to just not inherit any imports from the parent file.

But, as Erik just pointed out to me, *until* we have a concrete ambiguity, we shouldn't be saying what an incorrect program means. That's a job for recovery and error messaging. We should say what allowed programs mean, and we should not allow programs that we don't want the user to write.

As currently specified, a valid program can have `part id.id/id.id/id;` with no spaces between the `id`s and the `.` and `/`s.
If it doesn't follow that format, it's not a valid Dart program, and we don't *have* to say how it's wrong, just that it is wrong.

That does mean that `part of;` is valid and means the same as `part "package:of/of.dart";`.
This is not parsing recovery if we allow the syntax and then give a "file/package not found" error. (It's incredibly unlikely that `package:of/of.dart` is a part file, and the current file is inside `package:foo`'s `lib` directory, which is the only way it can not be an error.)

It's still surprising. It may be error-prone, if someone writes `part of;` and forgets to write a URI. Maybe they removed the `library foo;` because we tell them to, and then change `part of foo;` to just `part of;`. (It's obvious which file it's part of, right?)

If we want to prevent that, we can:
* Not allow unquited URIs in `part` directives, or
* Not consider a single identifier `of` as an unquoted URI in a `part` directive. (Basically the same as saying that `of` is a contextually reserved word, and *contextually* reserved words take precedence over everything else if they occur where we reserved them.)

Or we *can* do nothing, and have our tools recognize and give a tailored error message if someone does `part of;` and `package:of/of.dart` fails to exist or be a valid part of the current file. ("You wrote..., you probably meant to write ...".)